File: ruby/lib/baseh/checksum.rb
# frozen_string_literal: true

module Baseh
  # Version 1 checksum, spec section 6.2. Rolling polynomial over symbol
  # values, then modulus conversion into the checksum alphabet.
  module Checksum
    INITIAL_STATE = 17
    MULTIPLIER = 37

    module_function

    # Returns the checksum value in [0, modulus).
    # Spec 22: expandable generations may pass a shorter effective checksum
    # length; the modulus is then S^length instead of the profile default.
    def checksum_value(prepared, body, body_index = nil, length = prepared.checksum_length)
      body_index ||= BaseN.alphabet_index(prepared.body_alphabet)
      modulus =
        if length == prepared.checksum_length
          prepared.checksum_modulus
        else
          [prepared.checksum_alphabet.length, 1].max**length
        end

      state = INITIAL_STATE
      prepared.profile_id.each_byte do |byte|
        state = (state * MULTIPLIER + byte + 1) % modulus
      end
      state = (state * MULTIPLIER) % modulus
      body.each_char.each_with_index do |ch, pos|
        symbol_value = body_index[ch]
        unless symbol_value
          raise BasehError.new(
            "INVALID_CHARACTER",
            "Body symbol #{ch.inspect} is not in the body alphabet"
          )
        end
        state = (state * MULTIPLIER + symbol_value + pos + 1) % modulus
      end
      state
    end

    # Expected checksum string for a normalized body.
    def calculate_checksum(prepared, body, body_index = nil, length = prepared.checksum_length)
      return "" if length.zero?

      body_index ||= BaseN.alphabet_index(prepared.body_alphabet)
      value = checksum_value(prepared, body, body_index, length)
      BaseN.encode_base_n(value, prepared.checksum_alphabet, length)
    end
  end
end

File: python/src/baseh/checksum.py
"""Version 1 rolling polynomial checksum, spec section 6.2."""

from __future__ import annotations

from .basen import alphabet_index, encode_base_n
from .errors import INVALID_CHARACTER, BasehError
from .profile import PreparedProfile

_INITIAL_STATE = 17
_MULTIPLIER = 37


def checksum_value(
    profile: PreparedProfile,
    body: str,
    body_index: dict,
    checksum_length: int | None = None,
) -> int:
    """Return the checksum value in [0, modulus).
    Spec 22: expandable generations may pass a shorter effective checksum
    length; the modulus is then S^length instead of the profile default."""
    if checksum_length is None:
        checksum_length = profile.checksum_length
    if checksum_length == profile.checksum_length:
        modulus = profile.checksum_modulus
    else:
        base = len(profile.checksum_alphabet_norm) or 1
        modulus = base**checksum_length
    state = _INITIAL_STATE
    for byte in profile.profile_id.encode("ascii"):
        state = (state * _MULTIPLIER + byte + 1) % modulus
    state = (state * _MULTIPLIER) % modulus
    for pos, ch in enumerate(body):
        value = body_index.get(ch)
        if value is None:
            raise BasehError(
                INVALID_CHARACTER, f"Symbol {ch!r} is not in the body alphabet"
            )
        state = (state * _MULTIPLIER + value + pos + 1) % modulus
    return state


def calculate_checksum(
    profile: PreparedProfile, body: str, checksum_length: int | None = None
) -> str:
    """Compute the expected checksum string for a normalized body."""
    if checksum_length is None:
        checksum_length = profile.checksum_length
    if checksum_length == 0:
        return ""
    index = alphabet_index(profile.body_alphabet_norm)
    value = checksum_value(profile, body, index, checksum_length)
    return encode_base_n(value, profile.checksum_alphabet_norm, checksum_length)
